From 595eed49ef8cce8e63124e8168913e356a46671c Mon Sep 17 00:00:00 2001
From: Salvatore Bonaccorso <carnil@debian.org>
Date: Fri, 10 May 2024 09:05:53 -0400
Subject: [PATCH 10/10] junction: export-cache: cast to a type with a known
 size to ensure sprintf works

As reported in Debian, with the 64bit time_t transition for the armel
and armhf architecture, it was found that nfs-utils fails to compile
with:

	libtool: compile:
	export-cache.c: In function ?junction_flush_exports_cache?:
	export-cache.c:110:51: error: format ?%ld? expects argument of type ?long int?, but argument 4 has type ?time_t? {aka ?long long int?} [-Werror=format=]
	  110 |         snprintf(flushtime, sizeof(flushtime), "%ld\n", now);
	      |                                                 ~~^     ~~~
	      |                                                   |     |
	      |                                                   |     time_t {aka long long int}
	      |                                                   long int
	      |                                                 %lld

time_t is not guaranteed to be 64-bit, so it must be coerced into the expected
type for printf. Cast it to long long.

Reported-by: Vladimir Petko <vladimir.petko@canonical.com>
Closes: https://bugzilla.kernel.org/show_bug.cgi?id=218540
Link: https://bugs.debian.org/1067829
Link: https://bugs.launchpad.net/ubuntu/+source/nfs-utils/+bug/2055349
Fixes: 494d22396d3d ("Add LDAP-free version of libjunction to nfs-utils")
Suggested-by: Vladimir Petko <vladimir.petko@canonical.com>
Signed-off-by: Salvatore Bonaccorso <carnil@debian.org>
Signed-off-by: Steve Dickson <steved@redhat.com>
---
 support/junction/export-cache.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/support/junction/export-cache.c b/support/junction/export-cache.c
index 4e578c9b..00187c01 100644
--- a/support/junction/export-cache.c
+++ b/support/junction/export-cache.c
@@ -107,7 +107,7 @@ junction_flush_exports_cache(void)
 		xlog(D_GENERAL, "%s: time(3) failed", __func__);
 		return FEDFS_ERR_SVRFAULT;
 	}
-	snprintf(flushtime, sizeof(flushtime), "%ld\n", now);
+	snprintf(flushtime, sizeof(flushtime), "%lld\n", (long long)now);
 
 	for (i = 0; junction_proc_files[i] != NULL; i++) {
 		retval = junction_write_time(junction_proc_files[i], flushtime);
-- 
2.45.2

